### PR TITLE
Documented Unsafe.Unbox method

### LINKED ISSUE #2020: Unsafe.Unbox API needs warning about intended usage
Doc link: https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.unsafe.unbox
For full context, see conversation at https://github.com/dotnet/corefx/issues/34258.

The `Unsafe.Unbox<T>(object)` method is special because there are additional usage constraints developers must be aware of when calling this method, and the compiler won't necessarily inform the developer if they're about to violate these extra constraints. We instead need to make sure these are called out in this API's documentation.

In particular, the `unbox` IL instruction is intended to return a controlled-mutability managed pointer. Since we cannot represent this constraint via our API surface, the `Unsafe.Unbox<T>(object)` instruction instead returns a normal mutable `ref T`. However, developers __must not__ mutate the returned reference unless they are certain that `T` is a mutable struct type.

__Disallowed example:__

```cs
// 'int' is not a mutable struct type, so the below code is illegal
Unsafe.Unbox<int>(obj) = 30;
```

__Allowed example:__

```cs
// 'System.Drawing.Point' is a mutable struct type with public property setters, so the below code is legal
Unsafe.Unbox<System.Drawing.Point>(obj).X = 50;
```

For more information regarding allowed and disallowed operations on the return value of `Unsafe.Unbox<T>(object)`, see [ECMA-335 (PDF)](https://www.ecma-international.org/publications/files/ECMA-ST/ECMA-335.pdf), specifically § III.4.32 and III.1.8.1.2.2.

### PR DESCRIPTION
**Pull request by rpetrusha**
## Documented Unsafe.Unbox method

Fixes #2020 

/cc @GrabYourPitchforks

### CODE REVIEW

**Comment by GrabYourPitchforks** [comment]
I'd recommend giving a link back to the ECMA spec so that developers can see what "controlled-mutability managed pointer" really means, but otherwise LGTM. Thanks! :)

File: xml/System.Runtime.CompilerServices/Unsafe.xml [hunk 1]
```diff
@@ -1099,11 +1099,37 @@
...
+
+In contrast, a type such as <xref:System.Drawing.Point?<displayProperty=nameWithType> is a mutable struct with public property setters, so the following code is supported:
+
+```csharp
+Unsafe.Unbox<System.Drawing.Point>(obj).X = 50;
+```
+ 
+         ]]></format>
+         </remarks>       
+        <exception cref="T:System.NullReferenceException"><paramref name="obj" /> is <see langword="null" />, and <typeparamref name="T" /> is a non-nullable value type.</exception>
```
Comment (mairaw) [inline xml/System.Runtime.CompilerServices/Unsafe.xml]: Shouldn't this be box?

```suggestion
        <exception cref="T:System.NullReferenceException"><paramref name="box" /> is <see langword="null" />, and <typeparamref name="T" /> is a non-nullable value type.</exception>
```

File: xml/System.Runtime.CompilerServices/Unsafe.xml [hunk 2]
```diff
@@ -1099,11 +1099,37 @@
...
+         <exception cref="T:System.InvalidCastException"><paramref name="obj" /> is not a boxed value type.
```
Comment (mairaw) [inline xml/System.Runtime.CompilerServices/Unsafe.xml]: Ditto.

```suggestion
         <exception cref="T:System.InvalidCastException"><paramref name="box" /> is not a boxed value type.
```

File: xml/System.Runtime.CompilerServices/Unsafe.xml [hunk 3]
```diff
@@ -1099,11 +1099,37 @@
...
+         
+-or-
+
+<paramref name="obj" /> is not a boxed <typeparmref name="T" />.
```
Comment (mairaw) [inline xml/System.Runtime.CompilerServices/Unsafe.xml]: Ditto + bad tag.

```suggestion
<paramref name="box" /> is not a boxed <typeparamref name="T" />.
```

**Pull request merged**